File: rust/everybody-codes/src/bin/2025/y25d08.rs
use rayon::iter::ParallelIterator;
use everybody_codes::utilities::inputs::get_event_inputs;
use itertools::Itertools;
use rayon::iter::IntoParallelRefIterator;
use utilities::{minmax::minmax, parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Input<'a> = &'a str;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let (input1, input2, input3) = get_event_inputs(25, 8);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input1), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input2), stopwatch.lap().report());
    println!("3. {} ({})", part3(&input3), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse(input: Input) -> Vec<isize> {
    input.get_numbers::<isize>().map(|n| n - 1).collect()
}

fn part1(instructions: Input) -> usize {
    let half = 32 / 2;
    let instructions = parse(instructions);

    instructions.iter().tuple_windows()
        .filter(|&(a, b)| (a - b).abs() == half)
        .count()
}

fn part2(instructions: Input) -> usize {
    let instructions = parse(instructions);
    let crosses = instructions.iter().tuple_windows()
        .map(|(a, b)| {
            let (min, max) = minmax(a, b);
            (*min, *max)
        });

    let mut completed: Vec<(isize, isize)> = Vec::new();
    let mut knots = 0;

    for (a, b) in crosses {
        for &(ca, cb) in completed.iter() {
            let semicircle = ca + 1 .. cb;
            if semicircle.contains(&a) {
                // b must be outside
                if b < ca || b > cb {
                    knots += 1;
                }
            } else if semicircle.contains(&b) {
                if a < ca || a > cb {
                    knots += 1;
                }
            } 
        }
        completed.push((a, b));
    }
    knots
}

fn part3(instructions: Input) -> usize {
    let instructions = parse(instructions);
    let nails = 256;
    let crosses: Vec<_> = instructions.iter().tuple_windows()
        .map(|(a, b)| {
            let (min, max) = minmax(a, b);
            (*min, *max)
        })
        .collect();
    let combinations = (0..nails).tuple_combinations().collect::<Vec<(_, _)>>();
    combinations.par_iter()
        .map(|&(a, b)| {
            let mut cuts = 0;
            for &(ca, cb) in crosses.iter() {
                if a == ca && b == cb {
                    cuts += 1;
                    continue;
                }
                let semicircle = ca + 1 .. cb;
                if semicircle.contains(&a) {
                    // b must be outside
                    if b < ca || b > cb {
                        cuts += 1;
                    }
                } else if semicircle.contains(&b) {
                    if a < ca || a > cb {
                        cuts += 1;
                    }
                }
            }
            cuts
        })
        .max()
        .unwrap()
}

#[test]
fn default() {
    let (input1, input2, input3) = get_event_inputs(25, 8);
    assert_eq!(58, part1(&input1));
    assert_eq!(2924358, part2(&input2));
    assert_eq!(2792, part3(&input3));
}

// Input parsed (42μs)
// 1. 58 (6μs)
// 2. 2924358 (8.009ms)
// 3. 2792 (26.008ms)
// Total: 34.071ms

File: kotlin/everybody_codes/src/main/kotlin/org/gristle/pdxpuzzles/everybodycodes/y2025/Y25D08.kt
package org.gristle.pdxpuzzles.everybodycodes.y2025

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.runBlocking
import org.gristle.pdxpuzzles.everybodycodes.utilities.Day
import org.gristle.pdxpuzzles.utilities.combinatorics.getPairs
import org.gristle.pdxpuzzles.utilities.iteration.minMax
import org.gristle.pdxpuzzles.utilities.parsing.getInts
import kotlin.math.abs

object Y25D08 : Day {
    private fun parse(input: String): Sequence<Int> = input.getInts().map { it - 1 }
    private fun crosses(input: String): Sequence<Pair<Int, Int>> = parse(input)
        .windowed(2)
        .map { (a, b) -> minMax(a, b) }

    override fun part1(input: String): Int = parse(input)
        .windowed(2)
        .count { (a, b) -> abs(a - b) == 16 }

    override fun part2(input: String): Int {
        val completed = mutableListOf<Pair<Int, Int>>()
        var knots = 0

        for ((a, b) in crosses(input)) {
            for((ca, cb) in completed) {
                val semicircle = ca + 1 until cb
                val rng = ca..cb
                if (a in semicircle) {
                    if (b !in rng) {
                        knots++
                    }
                } else if (b in semicircle) {
                    if (a !in rng) {
                        knots++
                    }
                }
            }
            completed.add(a to b)
        }
        return knots
    }

    override fun part3(input: String): Int = runBlocking {
        val crosses = crosses(input).toList()
        val nails = 256
        val combinations = List(nails) { it }.getPairs()
        combinations
            .map { (a, b) ->
                async(context = Dispatchers.Default) {
                    var cuts = 0
                    for ((ca, cb) in crosses) {
                        if (a == ca && b == cb) {
                            cuts++
                            continue
                        }
                        val semicircle = ca + 1 until cb
                        val rng = ca..cb
                        if (a in semicircle) {
                            if (b !in rng) {
                                cuts++
                            }
                        } else if (b in semicircle) {
                            if (a !in rng) {
                                cuts++
                            }
                        }
                    }
                    cuts
                }
            }.awaitAll()
            .max()
    }
}

fun main() = Day.runDay(Y25D08::class)

//    Quest 1: 58 (1ms)
//    Quest 2: 2924358 (61ms)
//    Quest 3: 2792 (273ms)
//    Total time: 337ms